_posts/2021-12-14-残響.md

```diff
@@ -0,0 +1,65 @@
+
+# Re-Entrancy Attack
+This type of attack consist of calling recursively call a vulnerable function that didn't check it's own state. 
+This attack was the hearth of [The DAO](https://en.wikipedia.org/wiki/The_DAO_(organization)) Hack, an hacking attack that happened in June 2016 and drained the founds of the largest smart contract of that time, that held more 150 million dollars in Ether. A more detailed description of the history of the DAO's hack can be found [here](https://coinmarketcap.com/alexandria/article/a-history-of-the-dao-hack).
+
+# Exploit Creation
+To perform this kind of attack we can create a new contract [**Reentrancy.sol**](https://github.com/TowerofHanoi/towerofhanoi.github.io/tree/master/writeups_files/残響/Reentrancy.sol), in which we write the exploit.
+The contract must first call the existing Iterator.sol contract:
+```solidity
+    Iterator public iterator;
+
+    constructor(address payable etherIteratorAddress) public {
+        iterator = Iterator(etherIteratorAddress);
+    }
+```
+
+The heart of the exploit is the function [attack](https://github.com/TowerofHanoi/towerofhanoi.github.io/tree/master/writeups_files/残響/Reentrancy.sol#L22-L26): 
+```solidity
+function attack() external payable {
+    require(msg.value >= 1 ether);
+    iterator.donate.value(1 ether)(address(this)); 
+    iterator.withdraw(1 ether);
+} 
+```
+The attack function, because of the lack of state variables or function modifiers in the withdraw function of Iterator.sol, was able to call iterator.withdraw() multiple times before it finished executing. This resulted in more refunds and essentially recovering all of the Ether in the contract.
+
+# Transaction Sending
+After the creation of [Reentrancy.sol](https://github.com/TowerofHanoi/towerofhanoi.github.io/tree/master/writeups_files/残響/Reentrancy.sol), we connect to the blockchain of the challenge with the rpc endpoint given by `nc challs.xmas.htsp.ro 8014` after selecting _launch new instance_. With that command we also receive the Setup contract address and a uuid that will then be necessary to get the flag.
+Then by calling Setup.iterator we can obtain the address of the Iterator.sol contract that we can use to create the Reentrancy.sol contract by passing that address in its constructor.
```

Comment:
How do you call Setup.iterator and Reentrancy.attack? Do you have a python script? Commands? Please include them.

Reply:
Done.